File: rmi/RMIReaderWriter/src/Application/DataManager.java
package Application;

import java.io.FileNotFoundException;
import java.util.HashMap;
import java.util.concurrent.Semaphore;

public class DataManager {
	static HashMap<String, RWLockable> association;
  Semaphore mapMutex = new Semaphore(1);

	public static enum Strategy {
        FAVORING_READERS,
        FAVORING_WRITERS,
        SELFISH
    }

	public DataManager(String path, Strategy favoring) {
		association = new HashMap<>();
		association.put("arq1", LockerFactory(favoring, "arq1"));
		association.put("arq2", LockerFactory(favoring, "arq2"));
		association.put("arq3", LockerFactory(favoring, "arq3"));
	}

	public String read(String name) throws InterruptedException {
		String info = "";

    mapMutex.acquire();
		RWLockable lock = association.get(name);
    Resource r = lock.getResource();
    mapMutex.release();

		lock.requestRead();

		info = r.doRead();
		long threadId = Thread.currentThread().getId();
		System.out.println("Thread " + threadId + " lendo do arquivo " + name + ": " + info);

		lock.releaseRead();

		return info;
	}

	public void write(String path, int message) throws InterruptedException, FileNotFoundException {
    mapMutex.acquire();
    RWLockable lock = association.get(path);
		Resource r = lock.getResource();
    mapMutex.release();

		lock.requestWriting();

		r.doWrite(message);

		lock.releaseWriting();
	}

	private RWLockable LockerFactory(Strategy choice, String filepath) {
		switch(choice) {
		case FAVORING_READERS:
			return new RLocker(filepath);
		case FAVORING_WRITERS:
			return new WLocker(filepath);
		case SELFISH:
			return new NLocker(filepath);
		default:
			return new RLocker(filepath);
		}
	}
}

File: druby/ReaderWriter/Application/data_manager.rb
require_relative 'wlocker.rb'
require_relative 'rlocker.rb'

class DataManager
  class NoNameForFileError < StandardError; end

  def initialize(args)
    @@base ||= {}
    @@base[args[:filepath]] = locker_factory(args[:strategy], args[:filepath])
    @baseMutex = Mutex.new
  end

  def read(filepath, id)
    locker = nil
    @baseMutex.synchronize { locker = get_locker_with filepath }

    message = ""
    locker.access_read do |resource|
      message = resource.read id
    end
    message
  end

  def write(filepath, message, id)
    locker = nil
    @baseMutex.synchronize { locker = get_locker_with filepath }

    locker.access_write do |resource|
      resource.write(message, id)
    end
  end

  private
  def get_locker_with(filepath)
    @@base[filepath]
  end

  def locker_factory(strategy, filepath)
    case strategy
      when :favoring_readers
        RLocker.new(filepath)
      when :favoring_writers
        WLocker.new(filepath)
      when :no_favoring
        NLocker.new(filepath)
      else
        NLocker.new(filepath)
    end
  end
end
